Chunkwise image loader

lucas-diedrich opened the pull request:
## Description
This PR addresses the challenge that the currently implemented and planned image loaders require loading imaging data entirely into memory, typically as NumPy arrays. Given the large size of microscopy datasets, this is not always feasible.

To mitigate this issue, and as discussed with @LucaMarconato, this PR aims to introduce a generalizable approach for reading large microscopy files in chunks, enabling efficient handling of data that does not fit into memory.

Some related discussions. 
- https://github.com/scverse/spatialdata/issues/704#issue-2514386132
- https://github.com/scverse/spatialdata/issues/856#issue-2822806172
- https://github.com/lucas-diedrich/spatialdata-io/pull/1#issue-2697890163

## Strategy 
In this PR, we focus on `.tiff` images, as implemented in the `_tiff_to_chunks` function. 

1. Get a lazy representation of the image via a suitable reader function (here: `tifffile.memmap`)
2. Pre-define chunks that fit into memory, based on the dimensions of the image (`_compute_chunks`)
3.  Load small chunks via a custom reader function and pass the chunks to `dask.array` which is memory-mapped and avoids memory overflow (`_read_chunks`)
4. Reassembling the chunks into a `dask.array` (via `dask.array.block`)
5. Parse to Image2DModel.

The strategy is implemented in 
- `src/spatialdata_io/readers/generic.py` and
- `src/spatialdata_io/readers/_utils/_image.py`

## Future extensions
The strategy can be implemented for any image type, as long as it is possible to implement
1.  a lazy image-data loader 
2. define a custom reader function

We have implemented similar readers for openslide-compatible [whole slide images](https://github.com/lucas-diedrich/dvp-io/blob/e74b92778b95afd5c8272758015fc03f90937521/src/dvpio/read/image/openslide.py) and the [Carl-Zeiss microscopy](https://github.com/lucas-diedrich/dvp-io/blob/e74b92778b95afd5c8272758015fc03f90937521/src/dvpio/read/image/czi.py) format.

lucas-diedrich committed:
feature: Initial lazy tiff reader

lucas-diedrich committed:
Updated comments

lucas-diedrich committed:
Updated comments

lucas-diedrich committed:
Move utility functions to designated submodule readers._utils._image

lucas-diedrich committed:
Initial tests utils

lucas-diedrich committed:
Fixes edge cases for min coordinate

lucas-diedrich committed:
Added test for negative coordinates

lucas-diedrich committed:
Add support for png/jpg again

lucas-diedrich committed:
Add initial test

lucas-diedrich committed:
Fix: Fix jpeg and png reader, fix issues with local variable name

lucas-diedrich committed:
Merge branch 'main' of https://github.com/scverse/spatialdata-io into image-reader-chunkwise

melonora commented on src/spatialdata_io/readers/_utils/_image.py:
```diff
@@ -0,0 +1,129 @@
+    return positions, lengths
+
+
+def _compute_chunks(
+    dimensions: tuple[int, int],
+    chunk_size: tuple[int, int],
+    min_coordinates: tuple[int, int] = (0, 0),
+) -> NDArray[np.int_]:
+    """Create all chunk specs for a given image and chunk size.
+
+    Creates specifications (x, y, width, height) with (x, y) being the upper left corner
+    of chunks of size chunk_size. Chunks at the edges correspond to the remainder of
+    chunk size and dimensions
+
+    Parameters
+    ----------
+    dimensions : tuple[int, int]
```
```suggestion
def _compute_chunks(
    shape: tuple[int, int],
    chunk_size: tuple[int, int],
    min_coordinates: tuple[int, int] = (0, 0),
) -> NDArray[np.int_]:
    """Create all chunk specs for a given image and chunk size.

    Creates specifications (x, y, width, height) with (x, y) being the upper left corner
    of chunks of size chunk_size. Chunks at the edges correspond to the remainder of
    chunk size and dimensions

    Parameters
    ----------
    shape : tuple[int, int]
```
Just to stick to standard conventions:) Dimensions could be interpreted as `TCZYX`.

melonora commented on src/spatialdata_io/readers/_utils/_image.py:
```diff
@@ -0,0 +1,129 @@
+def _compute_chunk_sizes_positions(size: int, chunk: int, min_coord: int) -> tuple[NDArray[np.int_], NDArray[np.int_]]:
+    """Calculate chunk sizes and positions for a given dimension and chunk size"""
+    # All chunks have the same size except for the last one
+    positions = np.arange(min_coord, min_coord + size, chunk)
+    lengths = np.full_like(positions, chunk, dtype=int)
+
+    if positions[-1] + chunk > size + min_coord:
+        lengths[-1] = size + min_coord - positions[-1]
```
```suggestion
    positions = np.arange(min_coord, min_coord + size, chunk)
    lengths = np.minimum(chunk, min_coord + size - positions)
```
Think this is the equivalent two liner:) but just a bit nitpicky

lucas-diedrich replied:
Thanks, great suggestion :D, looked bit clunky before

melonora reviewed (approved):
Thanks for your contribution! I have 2 minor suggestions. I also saw that you use the `width` by `height` convention. Personally, I don't have a strong opinion here, though we could also stick to array api conventions. @LucaMarconato WDYT? Pre-approving for now.

melonora commented on src/spatialdata_io/readers/generic.py:
```diff
@@ -68,11 +81,57 @@ def geojson(input: Path, coordinate_system: str) -> GeoDataFrame:
+    list[list[DaskArray]]
+    """
+    # Lazy file reader
+    slide = tiffmmemap(input)
```
the tiff.memmap might not always work for example with compression or tiling so I would add a try, except clause.

lucas-diedrich replied:
Thanks, good catch! I added tests for compressed images and added, as suggested, a `try ... except` block to the generic reader function.

melonora reviewed (changes requested):
sorry had to change due to rethinking `memmap`. This does not always work, for example when dealing with compressed tiffs as far as I am aware.

lucas-diedrich committed:
Update src/spatialdata_io/readers/_utils/_image.py

Co-authored-by: Wouter-Michiel Vierdag <w-mv@hotmail.com>

lucas-diedrich committed:
[Refactor|API] Rename dimensions to shape to stick to numpy convention

lucas-diedrich committed:
[Refactor] Make suggested simplification of code, suggested by @melonora

lucas-diedrich committed:
[Test] Add test for compressed tiffs

lucas-diedrich committed:
[Fix] Account for compressed images

lucas-diedrich committed:
[Refactor] Remove unnecessary type hint

lucas-diedrich commented:
Thanks! Addressed your comments 
-Naming, simplification  https://github.com/scverse/spatialdata-io/pull/279/commits/46ed3e5f6e12f792b7f61da396a967ee87a22e6d, https://github.com/scverse/spatialdata-io/pull/279/commits/99731fedbc9502ddae055e9b26de7251ba5b52b7
- Try to read compressed images with `dask_image.imread` https://github.com/scverse/spatialdata-io/pull/279/commits/9e057de33bcdf5072bf3627220d05d6f7b57efde

LucaMarconato committed:
Merge branch 'main' into image-reader-chunkwise

LucaMarconato committed:
fix pre-commit

LucaMarconato committed:
fix transpose in image(); wip code review

LucaMarconato commented:
@lucas-diedrich yes still interested in it, I have time now to go through it. I added my "code review" in terms of "TODO" items for myself. I'll check the code and push some modifications if I see that some minor fixes are needed; I'll comment eventual larger changes, but the code looks good. I'll also do a benchmark with `asv` for performance.

LucaMarconato committed:
add test for dask-image fallback for compressed tiffs

LucaMarconato committed:
remove unused min_coordinate

LucaMarconato committed:
fix wrong dimension _compute_chunks(); cover with test

LucaMarconato committed:
np._int -> np.number

LucaMarconato committed:
fix indices in _read_chunks()

LucaMarconato committed:
better english

LucaMarconato committed:
better docstring

LucaMarconato committed:
wip benchmark (bugs)

lucas-diedrich committed:
[Test] Use small assymetric chunk sizes to capture any issues with the computation order of chunks/assembly

lucas-diedrich committed:
Add comment to clarify use of asymmetric chunk sizes in test_read_tiff

lucas-diedrich committed:
Follow standard convention of image dimensions throughout reader function. Switch axes order everywhere from (x, y) to (y, x)

lucas-diedrich committed:
Refactor chunk specification to follow (y, x) order and update related documentation

1. Enforce standard dimension order convention (y, x).

2. Change local variable names to better distinguish between chunk-level coordinates and pixel-level coordinates
- All chunk indices are indicated as such with the  prefix.
- The pixel-coordinates of individual chunks are now consistently named (y, x, height, width)

lucas-diedrich committed:
[Fix] Shape dimensions were inversed. Fix shape specification and make it self-documenting

lucas-diedrich committed:
chore: Remove Note and TODO

lucas-diedrich committed:
Clarify documentation

LucaMarconato committed:
fix pre-commit benchmark_image

LucaMarconato committed:
wip fix chunks

LucaMarconato committed:
improve chunks support

LucaMarconato committed:
benchmark for image() with synthetic data

LucaMarconato committed:
fix pre-commit

LucaMarconato commented:
Thanks @lucas-diedrich for spotting and fixing the bug, and for the new, cleaner code.

I also made some changes: have modified the `chunks` parameter to make it more general. With a single tuple of `(x, y)`, I could not pass `chunk` to `Image2DModel.parse()`, causing it being ignored for multiscale images. Now a more general chunking is implemented. The new cases are covered by tests.

I also finished writing the benchmark code (you can run it with `asv run --python=same -b IOBenchmarkImage --show-stderr -v`, here are the raw results when running it on my laptop, and the same results shown in a plot.

Takeaways:
- the benchmarks indicate consistent memory usage reduction as opposed to the old methods, both for single scale and multi scale images, in particular the memory usage is smaller than the case in which memmap is not used and the data is uncompressed
- the time execution is slightly worse than the case in which `memmap` is not used
- having the chunk size for `c` equal to the length of the `c` axis leads to less memory usage and a faster runtime than the case `c = 1`. This is expected because currently the assembly of chunks is implemented only for `x`, `y` and not for `c`. So a generalization of the code is worth.

<img width="1390" height="985" alt="image" src="https://github.com/user-attachments/assets/c0d30460-0644-4844-95ab-5f1df66e67d4" />

LucaMarconato committed:
c=1 vs c=3 benchmark
